### PR TITLE
Feature BI for Invoices

### LINKED ISSUE #31: BI for Invoices
Assuming we have a sales_engine that’s an instance of SalesEngine let’s initialize a SalesAnalyst like this:

```rb
sales_analyst = sales_engine.analyst
How many invoices does the average merchant have?
sales_analyst.average_invoices_per_merchant # => 10.49
sales_analyst.average_invoices_per_merchant_standard_deviation # => 3.29
```

Who are our top performing merchants?
Which merchants are more than two standard deviations above the mean?

```rb
sales_analyst.top_merchants_by_invoice_count # => [merchant, merchant, merchant]
```

Who are our lowest performing merchants?
Which merchants are more than two standard deviations below the mean?

```rb
sales_analyst.bottom_merchants_by_invoice_count # => [merchant, merchant, merchant]
```

Which days of the week see the most sales?
On which days are invoices created at more than one standard deviation above the mean?

```rb
sales_analyst.top_days_by_invoice_count # => ["Sunday", "Saturday"]
```

What percentage of invoices are not shipped?
What percentage of invoices are shipped vs pending vs returned? (takes symbol as argument)

```rb
sales_analyst.invoice_status(:pending) # => 29.55
sales_analyst.invoice_status(:shipped) # => 56.95
sales_analyst.invoice_status(:returned) # => 13.5
```
Effort point 5

### PR DESCRIPTION
**Pull request by NoahZinter**
Passes all tests and spec harness. 

If I run rubocop on this right now it does awful things to lines 25-34 of sales_analyst_mocks.rb. May be worth refactoring those lines anyway but rubo turns that into like 30 lines. 

- [x] Test Pass
- [x] Ready to Merge 

closes #31

### CODE REVIEW

File: lib/sales_analyst.rb [hunk 1]
```diff
@@ -58,9 +58,15 @@ def average_item_price_for_merchant(merchant_id)
     end
   end
 
+  def merchants_with_items
+    merchants_with_items = num_of_items_per_merchant.reject do |_merchant, num_of_items|
+      num_of_items.zero?
```
Comment (sami-p) [inline lib/sales_analyst.rb:63]: Awesome implementation of this new (to me) `.zero?` method!

File: lib/sales_analyst.rb [hunk 2]
```diff
@@ -80,11 +86,129 @@ def standard_deviation(values, mean)
...
+  def num_of_invoices_per_merchant
+    all_merchants.each_with_object({}) do |merchant, total_per_merchant|
+      total_invoices = all_invoices.count do |invoice|
+        invoice.merchant_id == merchant.id
+      end
+      total_per_merchant[merchant] = total_invoices
+    end
+  end
+
+  def average_invoices_per_merchant_standard_deviation
```
Comment (sami-p) [inline lib/sales_analyst.rb:103]: All looks great Noah! 🙌🏼 🎉 Am wondering if we can refactor this method name later to be a bit shorter?

File: lib/sales_analyst.rb [hunk 3]
```diff
@@ -80,11 +86,129 @@ def standard_deviation(values, mean)
...
+    mean = average_invoices_per_merchant
+    invoice_counts = num_of_invoices_per_merchant.values
+    std_dev = standard_deviation(invoice_counts, mean)
+  end
+
+  def top_merchants_by_invoice_count
+    mean = average_invoices_per_merchant
+    std_dev = average_invoices_per_merchant_standard_deviation
+
+    z = standard_deviations_of_mean(mean, std_dev, 2)
```
Comment (sami-p) [inline lib/sales_analyst.rb]: I remember in one of the lessons they advised us to steer away from using letters as variable names unless it relates directly to the material. Just a thought for refactoring! ✅

File: lib/sales_analyst.rb [hunk 4]
```diff
@@ -80,11 +86,129 @@ def standard_deviation(values, mean)
...
+    std_dev = average_invoices_per_merchant_standard_deviation
+
+    z = standard_deviations_of_mean(mean, std_dev, -2)
+
+    merchants = []
+    num_of_invoices_per_merchant.each_pair do |merchant, invoice_count|
+      merchants << merchant if invoice_count <= z
+    end
+    merchants
+  end
```
Comment (sami-p) [inline lib/sales_analyst.rb]: Amazing calculations!! Awesome job, Noah!!

File: lib/sales_analyst.rb [hunk 5]
```diff
@@ -80,11 +86,129 @@ def standard_deviation(values, mean)
...
+  def average_invoices_per_merchant
+    average = all_invoices.length.to_f / all_merchants.length
+    average.round(2)
+  end
+
+  def num_of_invoices_per_merchant
+    all_merchants.each_with_object({}) do |merchant, total_per_merchant|
+      total_invoices = all_invoices.count do |invoice|
+        invoice.merchant_id == merchant.id
+      end
```
Comment (mkrumholz) [inline lib/sales_analyst.rb:98]: ⛔️ Don't we have a  find_all_by_merchant_id method for invoices (invoice repo) already that you can use here @NoahZinter? -- then you could just .length on that?

File: spec/sales_analyst_mocks.rb [hunk 6]
```diff
@@ -1,3 +1,5 @@
+require './lib/item_repository'
+require './lib/invoice_repository'
```
Comment (sami-p) [inline spec/sales_analyst_mocks.rb:2]: We'll want to switch these two for alphabetical organization purposes

File: spec/sales_analyst_spec.rb [hunk 7]
```diff
@@ -2,6 +2,7 @@
 
 require './data/item_mocks'
 require './data/merchant_mocks'
+require './data/invoice_mocks'
```
Comment (sami-p) [inline spec/sales_analyst_spec.rb:5]: (Same as above: refactor alphabetically)

File: lib/sales_analyst.rb [hunk 8]
```diff
@@ -80,11 +86,129 @@ def standard_deviation(values, mean)
...
+    mean = average_invoices_per_merchant
+    std_dev = average_invoices_per_merchant_standard_deviation
+
+    z = standard_deviations_of_mean(mean, std_dev, 2)
+
+    merchants = []
+    num_of_invoices_per_merchant.each_pair do |merchant, invoice_count|
+      merchants << merchant if invoice_count >= z
+    end
+    merchants
```
Comment (mkrumholz) [inline lib/sales_analyst.rb]: ⛏ Could you refactor this method like #golden_items to make it more concise?

File: lib/sales_analyst.rb [hunk 9]
```diff
@@ -80,11 +86,129 @@ def standard_deviation(values, mean)
...
+    mean = average_invoices_per_merchant
+    std_dev = average_invoices_per_merchant_standard_deviation
+
+    z = standard_deviations_of_mean(mean, std_dev, -2)
+
+    merchants = []
+    num_of_invoices_per_merchant.each_pair do |merchant, invoice_count|
+      merchants << merchant if invoice_count <= z
+    end
+    merchants
```
Comment (mkrumholz) [inline lib/sales_analyst.rb]: Ditto the question from above

**Review by sami-p — APPROVED** [review]
This is some SERIOUS work 🙌🏼🙌🏼🙌🏼 Way to go Noah. Great job!

File: lib/sales_analyst.rb [hunk 10]
```diff
@@ -80,11 +86,129 @@ def standard_deviation(values, mean)
...
+  end
+
+  def invoice_created_at_times
+    if all_invoices[0].created_at.class == String
+      all_invoices.map do |invoice|
+        Time.parse(invoice.created_at)
+      end
```
Comment (mkrumholz) [inline lib/sales_analyst.rb:131]: 🤔 The created_at method in the invoices class should be able to handle this, rigtht? Curious about why you need this if/else syntax

File: lib/sales_analyst.rb [hunk 11]
```diff
@@ -80,11 +86,129 @@ def standard_deviation(values, mean)
...
+    end
+  end
+
+  def invoice_created_at_by_weekday
+    invoice_created_at_times.map do |time|
+      time.wday
+    end
+  end
+
+  def convert_wday_integers_to_hash
```
Comment (mkrumholz) [inline lib/sales_analyst.rb:148]: @NoahZinter this is the one I was looking at

File: lib/sales_analyst.rb (hunk 5 above)
Comment (NoahZinter) [reply lib/sales_analyst.rb:98]: Good catch! Thanks

File: lib/sales_analyst.rb (hunk 8 above)
Comment (NoahZinter) [reply lib/sales_analyst.rb]: I'll dig into that

File: lib/sales_analyst.rb (hunk 3 above)
Comment (NoahZinter) [reply lib/sales_analyst.rb]: Good catch!

File: spec/sales_analyst_mocks.rb (hunk 6 above)
Comment (NoahZinter) [reply spec/sales_analyst_mocks.rb:2]: ohh that's the alphabetizing we were talking about

File: lib/sales_analyst.rb [hunk 12]
```diff
@@ -80,11 +86,129 @@ def standard_deviation(values, mean)
...
+    weekday_hash.map do |weekday, invoices|
+      top_days << weekday if invoices > threshold
+    end
+    top_days
+  end
+
+  def invoice_status(status)
+    with_status = all_invoices.select do |invoice|
+      invoice.status == status
+    end.length
```
Comment (mkrumholz) [inline lib/sales_analyst.rb:177]: 🤔 Don't we have a #find_all_by_status method already in the InvoiceRepository class? We should use that, if so

File: lib/sales_analyst.rb (hunk 5 above)
Comment (NoahZinter) [reply lib/sales_analyst.rb:98]: Looks like we'll have to wait on that refactor until we have more access to repo methods from our inheritance refactor

File: lib/sales_analyst.rb (hunk 2 above)
Comment (mkrumholz) [reply lib/sales_analyst.rb:103]: Unfortunately that's from the spec so it has to stay this way

**Review by mkrumholz — APPROVED** [review]
I am good with merging this and dealing with the refactor after all the methods are working for my branch

**Pull request merged**